### ISSUE TITLE
OSError: Cannot understand given URI: <tf.Tensor 'filepath:0' shape=() dtype=string>.

### ISSUE THREAD

**Issue by jthibaut** [tensorflow/tensorflow#47891]
Dear All,
I am trying to implement a deep learning model using .tiff files
To read them I use skimage but I have difficulties to create a tensor flow dataset:
I did:
dataset_train = tf.data.Dataset.from_tensor_slices((X_train_path, y_train_path))
dataset_train = dataset_train.map(lambda x, y : [load_image(x), load_image(y)], num_parallel_calls=-1).batch(32)
Where:
def load_image(filepath):
    # read file
    im = skimage.io.imread(filepath)
    return im

and X_train_path is an object containing the list of the tif files

I have the following error message:
    OSError: Cannot understand given URI: <tf.Tensor 'filepath:0' shape=() dtype=string>.

Images are encoded on 4 channels.
Can you help me ?

**Comment by Saduf2019** [comment]
@jthibaut 

Can you please refer to similar issue and let u know: [link](https://stackoverflow.com/questions/57008293/use-tf-data-dataset-for-ppm-files)

**Comment by jthibaut** [comment]
Thanks
I already saw this solution but it does not work for .tiff files !

**Comment by Saduf2019** [comment]
@jthibaut 
Please share stand alone code such that we can replicate the issue reported such that we can replicate the issue, or if possible share a colab gist with the error.

**Comment by jthibaut** [comment]
Fine ! How can I do because I have a lot of pictures and I prefer to work on standalone rather than on collab as it is heavy each time to upload pictures and I have a limited workspace ?

**Comment by AdityaKane2001** [comment]
@jthibaut 
Use 
```python
def load_image(filepath):
    # read file
    im = skimage.io.imread(filepath.numpy())
    return im
```

**Comment by jthibaut** [comment]
If I do that I have the following message:
    AttributeError: 'Tensor' object has no attribute 'numpy'

**Comment by jthibaut** [comment]
Can you see it on collab ?
https://colab.research.google.com/drive/1H70CJKwaLNtxFWZDlw2vHZbFCapDlIz5#scrollTo=38owJ0189MbR

**Comment by AdityaKane2001** [comment]
@jthibaut 
I don't have access, can you share it as a gist?

**Comment by jthibaut** [comment]
I am sorry I am not proficient with github

https://github.com/jthibaut/Landscape-identification2/blob/main/MODEL1.ipynb

**Comment by AdityaKane2001** [comment]
no problem

**Comment by AdityaKane2001** [comment]
@jthibaut 
Ok, so you may try 2 things:
1. Remove the `@tf.function` decorator, as it executes in graph mode, and thus the values are not inferred using `.numpy()` function.
2. If that doesn't work, execute `tf.enable_eager_execution()` as soon as you import tensorflow.

Hope this helps.

**Comment by ShubhamShaswat** [comment]
@jthibaut 
Install tensorflow_io using `!pip install tensorflow_io` and then

Try this :
```python
import tensorflow_io as tfio

def load_image(filepath):
  img = tfio.experimental.image.decode_tiff(
   tf.io.read_file(filepath),
   index=0
  )
  return img
```

**Comment by jthibaut** [comment]
Thanks !
It seems to work but my data sets seems to be empty:
`dataset_train = tf.data.Dataset.from_tensor_slices((X_train_path, y_train_path))`
`dataset_train`
<TensorSliceDataset shapes: ((), ()), types: (tf.string, tf.string)>
`dataset_train = dataset_train.map(lambda x, y : [load_image(x), load_image(y)], num_parallel_calls=-1).batch(32)`
`dataset_train`
<BatchDataset shapes: ((None, None, None, None), (None, None, None, None)), types: (tf.uint8, tf.uint8)>

**Comment by ShubhamShaswat** [comment]
I am not sure about the reason for not showing the output shape, but the dataset itself is not empty. You can check that by running this code : 
```python
for ds in dataset_train.take(1):
 print(ds)
```

**Comment by jthibaut** [comment]
Thanks ! It work but I have a new problem: I used a U-net architecture and if I give as a loss function 'categorical_crossentropy' I have the following error message:
InvalidArgumentError:  logits and labels must be broadcastable: logits_size=[2097152,10] labels_size=[2097152,4]
	 [[node categorical_crossentropy/softmax_cross_entropy_with_logits (defined at <ipython-input-93-b64f662a0ca4>:1) ]] [Op:__inference_train_function_20000]

Function call stack:
train_function

I try to predict 10 classes of pictures. I have pictures with 4 channels. I don't konw if it explains the figures [2097152,10] and [2097152,4]. And I don't know how to check these parameters !

**Comment by ShubhamShaswat** [comment]
can you post the code for the same, meanwhile check if the labels shape that is your `y`  which should be same as your output shape that is `logits_size`

**Comment by jthibaut** [comment]
I put the link here:
https://colab.research.google.com/drive/1h63P4DhR3RJLu48IOSbgfgUz1eFYpFyj

**Comment by ShubhamShaswat** [comment]
I don't have access to this, when you share code through Colab select "Anyone with the link" instead of "Restricted" after clicking on "Get Link" option, this way people can see your code but can't edit.If you want, you can either choose "Viewer" , "Commenter" (they can only comment) or "Editor" ( can edit your code too). In your case,"viewer" seems fine too.

**Comment by jthibaut** [comment]
Does it work here ?
https://gist.github.com/jthibaut/d947b0956b0738f508163d7d876ee89e

**Comment by ShubhamShaswat** [comment]
Yes, I can see this one. Your outputs is of shape `(None, 256, 256, 10)`, the first one i.e. "None" is just your batch size. Now, In your case your dataset have two images as inputs, the second one is your label which is of shape `(256, 256, 4)`. Now you are using `sparse_categorical_loss` which is used when you have multiple classes for training. Now in here, you have two images that is your "input image" and "mask" . So using this loss function might not be the right choice.

Try changing these and see if it works for you : 

- try to get tweak your model to get the correct shape for your labels which is `mask` in your case
-  use `MSE` loss instead 

and if you can, it would be much better to explain what is you trying to do with your model. I have never worked with U-net.So I have a very limited idea about this architecture.

**Comment by jthibaut** [comment]
Yes, indeed ! I should have begun by that !
I just finished a training in data science and, in the frame of this training we had a project (taken from a kaggle competition) to perform but we (were 2) did not succeed to finish because we did not habe any help.
It was a cloud cell segmentation; try to find 4 shapes of clouds:
https://gist.github.com/jthibaut/8b8ceb4f5e1842a018d0ed017f8406e8

So, to train, I found a data challenge:
https://challengedata.ens.fr/participants/challenges/48/
https://github.com/earthcube-lab/challenge-ens/blob/master/README.md

The aim is to predict landscape cover (they are 10).
I thought to use a U-need model as it seemed to work for my cloud.
The problem here, is that the landscapes are not labelled.

So, may be, surely, my method is wrong at the beginning.
I don't know how to make appear the landscapes "classes". They are supposed to be in the mask files.
I did not want to use their file in order to train myself and also because I have a poor graphic card and no GPU.
So I want to make it simple, even though I have to limit the number of picture (that's what I did for the clouds because I took only pictures with one shape of cloud).
Regards
Thibaut

**Comment by jvishnuvardhan** [comment]
@jthibaut 
This is not Build/Installation or Bug/Performance issue. Please post this kind of support questions at [Stackoverflow](https://stackoverflow.com/questions/tagged/tensorflow). There is a big community to support and learn from your questions. GitHub is mainly for addressing bugs in installation and performance. Thanks!
